### PR TITLE
fix: limit parameter of schedule URI returns the same data

### LINKED ISSUE #16772: [BUG]: API returns the same next note for different deck id's
### Checked for duplicates?

- [X] This issue is not a duplicate

### Does it also happen in the desktop version?

- [ ] This bug does not occur in the latest version of Anki

### What are the steps to reproduce this bug?

- Retrieve the next scheduled card using `FlashCardsContract.ReviewInfo.CONTENT_URI` with `deckId` parameter
- Retrieve card with the returned info

### Expected behaviour

Return the next scheduled note id and card ordinal to the specific deck in the query

### Actual behaviour

Returns the same note Id for different decks

### Debug info

```text
AnkiDroid Version = 2.18.4 (f867da96b8304edaebd0361326316bc7514f8cde)

Backend Version = 0.1.38-anki24.04.1 (24.04.1 ccd9ca1a8309b80bcb50ddc5d99c7ce63440bce9)

Android Version = 12 (SDK 31)

ProductFlavor = play

Manufacturer = Xiaomi

Model = M2010J19SG

Hardware = qcom

Webview User Agent = Mozilla/5.0 (Linux; Android 12; M2010J19SG Build/SKQ1.211202.001; wv) AppleWebKit/537.36 (KHTML, like Gecko) Version/4.0 Chrome/126.0.6478.134 Mobile Safari/537.36

ACRA UUID = 5460b21c-2117-47fc-9cf3-908b997ffae9

FSRS Enabled = false

Crash Reports Enabled = true
```

## My code:
```kotlin
    fun queryNextCard(deck: DeckInfo): CardInfo? {
        println("Selected deck: ${deck.deckName} - ${deck.deckId}")

        val builder = FlashCardsContract.ReviewInfo.CONTENT_URI.buildUpon()
            .appendQueryParameter(
                FlashCardsContract.Note.DECK_ID_QUERY_PARAM, deck.deckId.toString()
            )
        val deckUri = builder.build()
        val deckCursor = AnkiPADroid.instance.applicationContext.contentResolver.query(
            deckUri, null, null, null
        )
        if (deckCursor == null || !deckCursor.moveToFirst())
            return null

        println("Deck Uri: $deckUri")

        val noteIdCol = deckCursor.getColumnIndex(FlashCardsContract.ReviewInfo.NOTE_ID)
        val cardOrdCol = deckCursor.getColumnIndex(FlashCardsContract.ReviewInfo.CARD_ORD)
        val noteId = deckCursor.getLong(noteIdCol)
        val cardOrd = deckCursor.getInt(cardOrdCol)

        deckCursor.close()

        val noteUri = Uri.withAppendedPath(
            FlashCardsContract.Note.CONTENT_URI, noteId.toString()
        )
        val cardsUri = Uri.withAppendedPath(noteUri, "cards")
        val specificCardUri = Uri.withAppendedPath(cardsUri, cardOrd.toString())

        val cardCursor = AnkiPADroid.instance.applicationContext.contentResolver.query(
            specificCardUri, null, null, null
        )
        if (cardCursor == null || !cardCursor.moveToFirst())
            return null


        println("Specific card Uri: $specificCardUri")
        val questionCol = cardCursor.getColumnIndex(FlashCardsContract.Card.QUESTION)
        val answerCol = cardCursor.getColumnIndex(FlashCardsContract.Card.ANSWER)

        val question = cardCursor.getString(questionCol)
        val answer = cardCursor.getString(answerCol)

        cardCursor.close()

        return CardInfo(noteId, cardOrd, question, answer)
    }
```

## Output:
```log
Selected deck: ankipadroid - 1716340404102
Deck Uri: content://com.ichi2.anki.flashcards/schedule?deckId=1716340404102
Specific card Uri: content://com.ichi2.anki.flashcards/notes/1674853668823/cards/0
天上的红星
天上的红星

<hr id=answer>

Selected deck: speaking - 1707769421598
Deck Uri: content://com.ichi2.anki.flashcards/schedule?deckId=1707769421598
Specific card Uri: content://com.ichi2.anki.flashcards/notes/1674853668823/cards/0
天上的红星
天上的红星

<hr id=answer>

Selected deck: pronunciation - 1674568133827
Deck Uri: content://com.ichi2.anki.flashcards/schedule?deckId=1674568133827
Specific card Uri: content://com.ichi2.anki.flashcards/notes/1674853668823/cards/0
天上的红星
天上的红星

<hr id=answer>

```

### Research

- [X] I am reporting a bug specific to AnkiDroid (Android app)
- [X] I have checked the [manual](https://ankidroid.org/docs/manual.html) and the [FAQ](https://github.com/ankidroid/Anki-Android/wiki/FAQ) and could not find a solution to my issue
- [ ] (Optional) I have confirmed the issue is not resolved in the latest alpha release ([instructions](https://docs.ankidroid.org/manual.html#betaTesting))

### PR DESCRIPTION
**Pull request by warleysr**
## Purpose / Description
I had some trouble integrating the AnkiDroid API in my app, mainly because some confusion on my part with the `CONTENT_URI` of `ReviewInfo` which `deckID` and `limit` aren't query parameters, not following the pattern of the others availables URIs. 

So I created the another URI to use the deck id as an appended path and passing limit as an optional query parameter, e.g. `/schedule/123456789?limit=5` leaving the existing one to not break any app that already uses it.

Also I found out that the `limit` parameter doesn't work in the existing schedule URI, returning the same card data N times and fixed this in this PR.

## Fixes
* Fixes #16772

## Approach
Creation of new URI in `CardContentProvider` and creation of `getQueuedCards()` function to solve the limit parameter issue.

## How Has This Been Tested?

Querying the `FlashCardsContract.ReviewInfo.CONTENT_URI` with appended path (new URI) and using selection arguments as the example provided in this file. The minimum working example I provided in the issue might help.

Used an Android 12 device.
 
## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/src/main/java/com/ichi2/libanki/sched/Scheduler.kt [hunk 1]
```diff
@@ -682,6 +682,14 @@ open class Scheduler(val col: Collection) {
     fun timeboxSecs(): Int {
         return col.config.get("timeLim") ?: 0
     }
+
+    fun getQueuedCards(limit: Int): List<Card> {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/libanki/sched/Scheduler.kt:686]: Hmm 🤔  - historically `libanki` is really only for things that are (or were) in the anki python library, there is a high bar to locating code here

Since this appears to be the only consumer of the limit-based query, could this code be inlined at the call site?

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt [hunk 2]
```diff
@@ -353,6 +355,23 @@ class CardContentProvider : ContentProvider() {
                 }
                 rv
             }
+            SCHEDULE_DECK -> {
+                val columns = projection ?: FlashCardsContract.ReviewInfo.DEFAULT_PROJECTION
+                val rv = MatrixCursor(columns, 1)
+                val selectedDeckBeforeQuery = col.decks.selected()
+                val deckId = uri.lastPathSegment!!.toLong()
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:351]: it appears that `rv` may be returned as a "short circuit" / "things didn't work" return value based on code below

Could we avoid the `!!` (which worries me every time I see it, no matter how improbable it is that it fail) and return rv if uri.lastPathSegment was for whatever reason null?

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt [hunk 3]
```diff
@@ -1254,6 +1243,23 @@ class CardContentProvider : ContentProvider() {
      * Calling package might be rogue if it has not declared #READ_WRITE_PERMISSION in its manifest */
     private fun knownRogueClient(): Boolean =
         !context!!.arePermissionsDefinedInManifest(callingPackage!!, FlashCardsContract.READ_WRITE_PERMISSION)
+
+    private fun fetchReviewInfo(col: Collection, limit: Int, rv: MatrixCursor, columns: Array<String>) {
```
Comment (mikehardy) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:1247]: Note to maintainers: I see `fetchReviewInfo` used in a previous commit, so this needs a squash-merge. Not a big deal, just noting it

**Review by mikehardy — CHANGES_REQUESTED** [review]
Hi there! 👋 

Always nice to see someone using the API - not too many consumers of it by count-of-apps but the apps that do can be quite popular. And PRs coming from API consumers are usually pretty great despite never seeing the authors before 😄  - like this one

I left a couple comments that in my mind are trivial. This works for me and I appreciate the care to be backwards compatible, that certainly helps with the decision to integrate or not

**Comment by mikehardy** [comment]
Curious if your git commit hook didn't run? Looks like the ktlint autoformatter would fix this, and should fix it when you attempt to commit? Just a trivial spacing issue is stopping lint from passing:

```

> Task :AnkiDroid:ktlintMainSourceSetCheck FAILED
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/main/java/com/ichi2/libanki/sched/Scheduler.kt:692:1 Unexpected blank line(s) before "}"
```

**Comment by mikehardy** [comment]
I see you and David have interacted a fair bit on the linked issue so I'm flagging him for review as well. I'm happy to be overruled on any of my review comments if David has other ideas

**Review by mikehardy — APPROVED** [review]
simply incorporates all my suggestions so of course I'm +1 but as mentioned I remain open to override by anything David thinks since you two interacted already

Fantastic, IMHO though. Cheers

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt [hunk 4]
```diff
@@ -334,25 +336,31 @@ class CardContentProvider : ContentProvider() {
...
                     col.decks.select(selectedDeckBeforeQuery)
                 }
                 rv
             }
+            SCHEDULE_DECK -> {
+                val columns = projection ?: FlashCardsContract.ReviewInfo.DEFAULT_PROJECTION
+                val rv = MatrixCursor(columns, 1)
+                val selectedDeckBeforeQuery = col.decks.selected()
+                val deckId = uri.lastPathSegment?.toLong() ?: return rv
+                val limit = uri.getQueryParameter("limit")?.toInt() ?: 1
```
Comment (david-allison) [inline AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:352]: (disclaimer: I haven't done much API work)

As far as I understand it, query parameters (after the `?`) are not expected within Content Provider URIs.

I suspect this was due to URL length limits, and has been de-facto standardized as such

https://developer.android.com/guide/topics/providers/content-provider-basics#ContentURIs

The documentation does not comment on, or use parameters. It feels that `selection` should be used for filtering, as we currently do in `NOTES` and `NOTES_V2`.

----

Inconsistency: our `bulkInsert` method does a query parameter, I believe this inconsistency is because Android's `bulkInsert` does not expose alternate arguments.

**Review by david-allison — CHANGES_REQUESTED** [review]
Code is really high quality! Thanks so much.

Please bear in mind that this code was written before my time and I've had limited interactions with it.

That being said: it's our public API, so I'm being much more nitpicky than normal

----

Could you add test(s) to `ContentProviderTest`:

https://github.com/ankidroid/Anki-Android/blob/06e8ac2ef3e0d35fb3efc59aabb38c796b276b04/AnkiDroid/src/androidTest/java/com/ichi2/anki/tests/ContentProviderTest.kt#L74

----

## API Design

TL;DR: I don't think we should include the new endpoint

We are already somewhat unusual in our `ContentProvider` design

* `limit` should be a selection parameter, not a query parameter, justification in a comment below

---

I'm a mild oppose to adding the new endpoint on grounds of consistency (I'm undecided here)

* Our current endpoints are 'list/detail':
    * `decks/` lists all decks
    * `decks/1` lists a deck
    * `notes/` lists all notes
    *  `notes/1` lists a note

The parameter changes the semantics of the request.

We don't have this list/detail distinction with `schedule`. 

`/schedule` lists the schedule for the currently selected deck
`/schedule/1` lists the schedule for deck `X`

`DeckId` doesn't change the semantics of the request here, so should be a selection argument.

File: AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt (hunk 4 above)
Comment (mikehardy) [reply AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:352]: Okay, sot hat's the criticism, and that seems well-reasoned. So the concrete proposal is:

- maintain current endpoint of /schedule ?
- add selection criteria of deck
- add selection criteria of limit (this is subtle now, since Android 11 and changed ContentProvider limit behavior)

So the underlying code would move the new logic into the schedule endpoint, and see what selection criteria were given - if deckid was not given, execute the old code to get data, if deckid was given execute the new code here

And then (the fancier bit...) if android O+ see if limit was sent in the bundle, if < O see if limit was sent in selection query as well and if so get it from query args. Either way if a limit was sent in, restrict results to limit ? (https://stackoverflow.com/a/63551176/9910298)

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:352]: I don't believe the linked SO post is necessary - we're in control of the provider (as the server, rather than the client) and can handle a limit provided from a client so we just serve that amount of data

----

Steps would be:

* Test `/schedule` with DeckId/limit (with bug fixed) 
* Merge

Comment (warleysr) [reply AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:352]: > Inconsistency: our `bulkInsert` method does a query parameter, I believe this inconsistency is because Android's `bulkInsert` does not expose alternate arguments.

I agree with this, query parameters aren't really used in any other endpoint, but following the other endpoints that I mentioned, the correct standardization would be `/schedule/<deckID>/<limit>` ?

Comment (david-allison) [reply AnkiDroid/src/main/java/com/ichi2/anki/provider/CardContentProvider.kt:352]: > Many providers let you access a single row in a table by appending an ID value to the end of the URI. For example, to retrieve a row whose _ID is 4 from the User Dictionary Provider, you can use this content URI:

https://developer.android.com/guide/topics/providers/content-provider-basics#ContentURIs

This is what we're doing with other endpoints

As above, this changes the meaning of the endpoint from "many" to "one", and therefore makes sense to be included in the URL

----

`ContentProvider` expects `selection` (/`sectionArgs`) to be used for selections, and we should follow this standardization to the best of our ability

`/schedule/<deckID>/<limit>` is:

* ambiguous (no expectation of what the second parameter could be. `deckId` is arguably the primary parameter for `schedule`, but it's not as obvious as `notes/1` referring to the note with ID 1)
* unmaintainable: the URL endpoint gets more complex for each parameter added
* unmaintainable: over time, we will reach the limit of a URL

It does not follow the standardization of the other endpoints: neither of the parameters change the shape/semantics of the call/results

**Review by david-allison — APPROVED** [review]
Great stuff, cheers!

If you write a quick comment detailing where our documentation is lacking regarding the parameters, I'll put it up as a documentation issue
